dix/req.c

```diff
@@ -1771,35 +1772,57 @@ static int dix_cas_rops_send(struct m0_dix_req *req)
 		cctg_id.ci_layout.dl_type = layout->dl_type;
 		rc = m0_dix_ldesc_copy(&cctg_id.ci_layout.u.dl_desc,
 				       &layout->u.dl_desc);
-		M0_LOG(M0_DEBUG, "Processing dix_req %p[%u] "FID_F
-				  " creq=%p "FID_F,
-				  req, req->dr_type,
-				  FID_P(&req->dr_indices[0].dd_fid),
-				  creq, FID_P(&cctg_id.ci_fid));
+		if (rc == 0) {
+			M0_LOG(M0_DEBUG, "Processing dix_req %p[%u] "FID_F
+			       " creq=%p "FID_F,
+			       req, req->dr_type,
+			       FID_P(&req->dr_indices[0].dd_fid),
+			       creq, FID_P(&cctg_id.ci_fid));
 
-		switch (req->dr_type) {
-		case DIX_GET:
-			rc = m0_cas_get(creq, &cctg_id, &cas_rop->crp_keys);
-			break;
-		case DIX_PUT:
-			rc = m0_cas_put(creq, &cctg_id, &cas_rop->crp_keys,
-					&cas_rop->crp_vals, req->dr_dtx,
-					cas_rop->crp_flags);
-			break;
-		case DIX_DEL:
-			rc = m0_cas_del(creq, &cctg_id, &cas_rop->crp_keys,
-					req->dr_dtx, cas_rop->crp_flags);
-			break;
-		case DIX_NEXT:
-			rc = m0_cas_next(creq, &cctg_id, &cas_rop->crp_keys,
-					 req->dr_recs_nr,
-					 cas_rop->crp_flags | COF_SLANT);
-			break;
-		default:
-			M0_IMPOSSIBLE("Unknown req type %u", req->dr_type);
+			switch (req->dr_type) {
+			case DIX_GET:
+				rc = m0_cas_get(creq, &cctg_id,
+						&cas_rop->crp_keys);
+				break;
+			case DIX_PUT:
+				rc = m0_cas_put(creq, &cctg_id,
+						&cas_rop->crp_keys,
+						&cas_rop->crp_vals,
+						req->dr_dtx,
+						cas_rop->crp_flags);
+				break;
+			case DIX_DEL:
+				rc = m0_cas_del(creq, &cctg_id,
+						&cas_rop->crp_keys,
+						req->dr_dtx,
+						cas_rop->crp_flags);
+				break;
+			case DIX_NEXT:
+				rc = m0_cas_next(creq, &cctg_id,
+						 &cas_rop->crp_keys,
+						 req->dr_recs_nr,
+						 cas_rop->crp_flags |
+						 COF_SLANT);
+				break;
+			default:
+				M0_IMPOSSIBLE("Unknown req type %u",
+					      req->dr_type);
+			}
+			m0_cas_id_fini(&cctg_id);
 		}
-		m0_cas_id_fini(&cctg_id);
+
 		if (rc != 0) {
+			/*
+			 * Treat failed and not sent CAS requests as executed
```

Comment:
How many of such failed requests can be treated as executed? What if all of them are failed?

Reply:
If all are failed then -EFAULT will be returned
```
	if (rop->dg_cas_reqs_nr == 0)
		return M0_ERR(-EFAULT);
```

Reply:
Regarding allowed maximum number of TRANSIENT failures - as I know it is under discussion now.